### PR TITLE
Experimental support for font hinting

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
Traditionally, RCT2 has used a sprite-based font to render its strings. Since OpenRCT2's inception, support has been added for rendering with *TrueType* fonts to accomodate languages that do not use a Western script, such as Chinese, Korean, and Japanese.

In the past, OpenRCT2's palette has been a point of frustration, as it prevented implementation of direct font hinting ('anti-aliasing'), leading the Korean community to raise their minimum font size from 8 to 11 points (cf. #2317). This is not an ideal solution, however, as it leads text strings to overflow from their containing UI elements, since widgets are currently not flexible.

This PR, while a work in progress, is a crude attempt of implementing support for font hinting in OpenRCT2. Let me start off with some examples.

## Screenshots

(Before on the left, after on the right.)

scenario_ru_before scenario_ru_after

settings_jp_before settings_jp_after

settings_ru_before settings_ru_after

## Known issues

The current implementation does not play nicely with all window background colours — in some cases even overflowing into the wrong colour family entirely. I am hopeful we can resolve this as my understanding of the palette system deepens, however.

I have been eyeing `ColourMapA` in colour.c — it seems to be exactly what is needed. One caveat: in some cases a darker shade is needed, while the others require a lighter one. Outlined texts, like in window titles, are probably related here.

## In conclusion

Ideas or other contributions would be most appreciated! I'm stuck right now, but excited for the possibilities, so I thought I'd put it the code out here for now.

### CODE REVIEW

**Comment by janisozaur** [comment]
We already have `[font]` section in `config.ini` describing various font properties, can the shaded v. solid option be included there as well?

**Comment by Gymnasiast** [comment]
And if you add it, probably best to turn it on by default (assuming it won't be used for the sprite font).

**Comment by marijnvdwerf** [comment]
Maybe we could create a function (with a cache) that calculates the blend between two colours. And then gets the closest palette index based on the closest colour.

**Comment by AaronVanGeffen** [comment]
@janisozaur Making it optional was fairly straightforward (da4dae6cbd3fec19d8ee8c42e66d9ea95542b467), although I'm not too keen on the added compile-time dependency. Perhaps it's no big deal, though?

@marijnvdwerf Hmm, I was hoping to avoid blending altogether, but rather work from the background colour. Depending on the intensity of the foreground colour, we'd either use a darker or lighter shade. Would clamping on the colour palette in such a fashion be feasible?

**Comment by AaronVanGeffen** [comment]
Getting there. Just to give an update, things look like this as of 6130d74e3972c276ea903567ad984430fad1828e:

screen shot 2017-09-24 at 01 33 43

**Comment by AaronVanGeffen** [comment]
Getting there. Here's what things look like as of e3091079ed5c3145ee8be26da9434bef31319142:

screen shot 2017-09-24 at 01 44 00

**Comment by janisozaur** [comment]
@telk5093 @YJSoft @daihakken @izhangfei can you guys check how it works for you?

**Comment by YJSoft** [comment]
image

Looks OK except IME part(where background is black)

tested with below config
```
[font]
file_name = "NanumGothic.ttf"
font_name = "NanumGothic"
x_offset = 0
y_offset = -2
size_tiny = 9
size_small = 11
size_medium = 12
size_big = 13
height_tiny = 14
height_small = 14
height_medium = 14
height_big = 20
enable_hinting = true
```

**Comment by AaronVanGeffen** [comment]
Thank you for testing, @YJSoft. I'm sure at one point the IME looked alright, so I should be able to fix it.

**Comment by Gymnasiast** [comment]
@AaronVanGeffen Any update on fixing the IME?

**Comment by AaronVanGeffen** [comment]
I had a quick look the other day, and while I'm sure I can fix it, it's a little less trivial than I assumed. Regrettably, I'm a little busy this week, but I'm hopeful can have another go at it this weekend.

**Comment by AaronVanGeffen** [comment]
The IME issue should be fixed with 4bff727 adding support for black backgrounds. @YJSoft, could you confirm?

While doing some more thorough testing myself (sparking #6372, since merged), I encountered two more rare cases: the bright red text displaying losses in the finances window, and the light sea green colour in the theme window. Regrettably, I've had to add exceptions for both of them.

I think this is ready for review and testing, once more.

File: src/openrct2/drawing/string.c [hunk 1]
```diff
@@ -812,6 +812,12 @@ static void ttf_draw_string_raw_ttf(rct_drawpixelinfo *dpi, const utf8 *text, te
                             } else if (*dst < 10) {
                                 // For black backgrounds, hint using grey pixels.
                                 *dst = 13;
+                            } else if (colour == 153) {
```
Comment (Gymnasiast) [inline src/openrct2/drawing/string.c]: Is there no constant for this? I remember @marijnvdwerf replacing all instances.

Comment (AaronVanGeffen) [reply src/openrct2/drawing/string.c]: `interface/colour.h` has a few colour constants, but not for any of these. Please correct me if I'm overlooking any, of course.

Comment (Gymnasiast) [reply src/openrct2/drawing/string.c]: I think you need to add the appropriate PALETTE_INDEX_x entries in the enum, but I'd like to hear from @marijnvdwerf first.

Comment (marijnvdwerf) [reply src/openrct2/drawing/string.c]: Either use `ColourMap`, or add a new `PALETTE_INDEX` entry.

Comment (AaronVanGeffen) [reply src/openrct2/drawing/string.c]: I've applied `ColourMapA` for now, where possible.  Suggestions for the other cases are welcome, too... I tried fiddling with the `BASE_COLOUR` macro as well, but I'm pressed for time and it wasn't quite working out yet, so I've left that out for now.

**Comment by YJSoft** [comment]
screenshot 2017-10-07 07 02 48

some letters are still hard to read, but I think it's font file's problem(letter too complex for small sizes)

**Comment by AaronVanGeffen** [comment]
Thanks for testing! You're right that the glyphs are a little complicated for this font size, but there may be something we can do still — I have an idea. Could you check back here in, say, 15 minutes?

**Comment by AaronVanGeffen** [comment]
@YJSoft Does abe79b7a937bed427ae285d33f0ffbe6fd006773 make things a little better for you? (It doesn't seem to make much of a difference when I test Japanese, but it's worth a shot...)

**Comment by YJSoft** [comment]
screenshot 2017-10-07 07 49 00
much better

**Comment by AaronVanGeffen** [comment]
Thanks again! It still doesn't look ideal, but at least the strokes are all visible, I guess.

@Gymnasiast, @marijnvdwerf Any more suggestions for improvement?

File: src/openrct2/drawing/scrolling_text.c [hunk 2]
```diff
@@ -1567,7 +1567,9 @@ void scrolling_text_set_bitmap_for_ttf(utf8 *text, sint32 scroll, uint8 *bitmap,
                 uint8 *dst = &bitmap[scrollPosition];
 
                 for (sint32 y = 0; y < height; y++) {
-                    if (src[y * pitch + x] != 0) *dst = colour;
+                    if (src[y * pitch + x] > 92 || (src[y * pitch + x] != 0 && !gConfigFonts.enable_hinting)) {
```
Comment (Gymnasiast) [inline src/openrct2/drawing/scrolling_text.c]: Brace placement

File: src/openrct2/drawing/ttf_sdlport.c [hunk 3]
```diff
@@ -1261,6 +1281,145 @@ TTFSurface *TTF_RenderUTF8_Solid(TTF_Font *font,
...
+    textbuf->h = height;
+    textbuf->pitch = width;
+    textbuf->pixels = calloc(1, width * height);
+
+    /* Adding bound checking to avoid all kinds of memory corruption errors
+       that may occur. */
+    dst_check = (uint8*)textbuf->pixels + textbuf->pitch * textbuf->h;
+
+    /* Fill the palette with NUM_GRAYS levels of shading from bg to fg */
+    /*
```
Comment (Gymnasiast) [inline src/openrct2/drawing/ttf_sdlport.c]: Commented-out code; should be removed.

File: src/openrct2/drawing/ttf_sdlport.c [hunk 4]
```diff
@@ -1261,6 +1281,145 @@ TTFSurface *TTF_RenderUTF8_Solid(TTF_Font *font,
...
+
+    for ( index = 0; index < NUM_GRAYS; ++index ) {
+        palette->colors[index].r = bg.r + (index*rdiff) / (NUM_GRAYS-1);
+        palette->colors[index].g = bg.g + (index*gdiff) / (NUM_GRAYS-1);
+        palette->colors[index].b = bg.b + (index*bdiff) / (NUM_GRAYS-1);
+    }
+    */
+
+    /* check kerning */
+    use_kerning = FT_HAS_KERNING( font->face ) && font->kerning;
```
Comment (Gymnasiast) [inline src/openrct2/drawing/ttf_sdlport.c]: `FT_HAS_KERNING(font->face)`

**Comment by Gymnasiast** [comment]
The formatting needs a few kicks.

**Comment by AaronVanGeffen** [comment]
The formatting concerns have been addressed per 86f28979c5b8bfd1c8d89288de37ba53cb70f4ff. I addressed each of the commits individually; three commits have been squashed into others to remove some experimental values. Effectively keeping the end results, while preserving the history of this PR over time.

I have not added a line to the changelog yet, as I expect merge conflicts there. Give me a sign if you think this is ready for merge, and I'll happily add it.

**Comment by AaronVanGeffen** [comment]
Your formatting concerns have been addressed per 822286afc0816f986fec93eae20ebbe22be7b373. I have addressed each of the commits individually; three commits have been squashed into others to remove some experimental values. Effectively keeping the end results, while preserving the history of this PR over time.

I have not added a line to the changelog yet, as I expect merge conflicts there. Give me a sign if you think this is ready for merge, and I will happily add it.

File: src/openrct2/drawing/scrolling_text.c [hunk 5]
```diff
@@ -1566,8 +1566,12 @@ void scrolling_text_set_bitmap_for_ttf(utf8 *text, sint32 scroll, uint8 *bitmap,
             if (scrollPosition > -1) {
                 uint8 *dst = &bitmap[scrollPosition];
 
-                for (sint32 y = 0; y < height; y++) {
-                    if (src[y * pitch + x] != 0) *dst = colour;
+                for (sint32 y = 0; y < height; y++)
+                {
+                    if (src[y * pitch + x] > 92 || (src[y * pitch + x] != 0 && !gConfigFonts.enable_hinting))
```
Comment (marijnvdwerf) [inline src/openrct2/drawing/scrolling_text.c]: What is `92`?

File: src/openrct2/drawing/string.c [hunk 6]
```diff
@@ -798,7 +799,44 @@ static void ttf_draw_string_raw_ttf(rct_drawpixelinfo *dpi, const utf8 *text, te
...
+                            else if (colour == ColourMapA[COLOUR_BRIGHT_RED].light)
+                            {
+                                // Exception: losses in finance window.
+                                *dst = ColourMapA[COLOUR_BRIGHT_RED].lightest;
+                            }
+                            else
+                            {
+                                // For other texts, hinting direction depends on the intensity of the
+                                // foreground colour.
+                                bool intensive = colour % 12 > 4 && colour % 12 < 10;
```
Comment (marijnvdwerf) [inline src/openrct2/drawing/string.c]: Do these numbers mean anything, or are they the result of experimentation?

File: src/openrct2/drawing/ttf_sdlport.c [hunk 7]
```diff
@@ -1261,6 +1282,146 @@ TTFSurface *TTF_RenderUTF8_Solid(TTF_Font *font,
     return textbuf;
 }
 
+TTFSurface *TTF_RenderUTF8_Shaded(TTF_Font *font,
+                const char *text, uint32 fg, uint32 bg)
+{
+    bool first;
```
Comment (marijnvdwerf) [inline src/openrct2/drawing/ttf_sdlport.c]: @IntelOrca Do we do variable predeclaration if the files come from an external source?

Comment (IntelOrca) [reply src/openrct2/drawing/ttf_sdlport.c]: I don't know.

File: src/openrct2/drawing/string.c (hunk 6 above)
Comment (AaronVanGeffen) [reply src/openrct2/drawing/string.c]: `colour % 12` gets you the palette row in question, as you probably figured. I found the domain between 4 and 10 through experimenting — these seem to work reasonably well across the board, save for the few exceptions as implemented.

File: src/openrct2/drawing/ttf_sdlport.c (hunk 7 above)
Comment (AaronVanGeffen) [reply src/openrct2/drawing/ttf_sdlport.c]: Save for @Gymnasiast's comments on formatting, I stuck to the implementation of the counterparts that were already there — which did leave the predeclaration intact.

Personally, I think it might come in handy not to refactor the SDL2_ttf code too much, if we ever want to merge changes from upstream.

File: src/openrct2/drawing/scrolling_text.c (hunk 5 above)
Comment (AaronVanGeffen) [reply src/openrct2/drawing/scrolling_text.c]: Here, it's an arbitrary threshold for when to interpret a greyscale pixel as a monochrome one for the banners. As my changes change the rendering function to work on the domain [0, 255] rather than [0, 1], a threshold is needed to stick with a monochrome palette for the banners.

I think I fiddled around with it in steps of 8 or 12. It could probably be 90 or 95 as well; we're obviously not bound by RCT2's palette as far as the `src` bitmap is concerned...

File: src/openrct2/drawing/ttf_sdlport.c (hunk 7 above)
Comment (Gymnasiast) [reply src/openrct2/drawing/ttf_sdlport.c]: I think predeclarations are fine in some cases. This is one.

**Comment by AaronVanGeffen** [comment]
@marijnvdwerf Cheers for the in-depth review. You're right to be weary of arbitrary thresholds and the likes, of course. On Gitter, you mentioned you were working on proper colour blending to replace the approach I've taken in this PR. That's something I most welcome, of course. If it's close to being finished, I'll happily incorporate it into this PR as well, dropping my 'hackier' commits. If not, then perhaps you could include it in #6075 instead, for both our sanities.

File: src/openrct2/drawing/string.c [hunk 8]
```diff
@@ -739,7 +739,7 @@ uint8 findClosestIndex(uint8 red, uint8 green, uint8 blue)
     return closest;
 }
 
-uint8 blend(const uint8 base, const uint8 layer)
```
Comment (marijnvdwerf) [inline src/openrct2/drawing/string.c]: Naming could probably be improved upon. As well as pointing out that the values are palette indexes.

Comment (AaronVanGeffen) [reply src/openrct2/drawing/string.c]: Agreed, it's probably not clear what the purpose of a 'blend' function in string.c. I was thinking it might be better to export these functions from colour.{c,h} instead, and move the blend table there as well. What are your thoughts on that?

**Comment by AaronVanGeffen** [comment]
The new blending code introduced by @marijnvdwerf in 799e202bec621fba375c682d19697f948b6f46fa works quite well overall, and is generally an improvement, especially conceptually. However, it did not account for the remappable colours in the palette; I've introduced `isRemappableColour` in b4941b9a56ff47aa9a7d37869ee57c22a71f38ae to address this to counter flickering text.

There is one more regression I think should be addressed before merging. I'm not sure how to solve it yet, however. Consider the scenario window on 1c5623fe091984b795316328bd5ba2e770843dd5:

screenshot_20171013_203900

Here's the same window on b4941b9a56ff47aa9a7d37869ee57c22a71f38ae:

screenshot_20171013_202659

As you can see, the glyphs are rendered much more thinly, making them harder to read.

**Comment by marijnvdwerf** [comment]
image
The only dynamic palette is the stuff following and including 0xE6 (360)

**Comment by AaronVanGeffen** [comment]
Ah, I see. I was going by the [palette spreadsheet](https://docs.google.com/spreadsheets/d/1OtTAytgtecqW7bbN8K6lb6dw4LYUMiXXBb7ZLGyk__U), but I considered all remapping areas rather than just the ones that are shifted dynamically. Will adjust the relevant commit.

**Comment by marijnvdwerf** [comment]
Tofu renders pretty badly:
image

**Comment by AaronVanGeffen** [comment]
Hmm, that tofu renders badly is unfortunate, but I wonder whether it's worth it overall. I sent this over Gitter earlier, but here's a screenshot comparing the 'mono' and 'light' hinting modes; 'normal' produces similar results to 'mono'.

comparison

As you can see, especially the more complicated glyphs are rendered much more clearly, while they are still adequately hinted. Hence, for Japanese, I would strongly prefer 'light' over 'mono'.

**Comment by AaronVanGeffen** [comment]
@marijnvdwerf I think I've fixed it. As it turns out, the tofu rendering was fine on Linux, but it did occur consistently on macOS. While I'd rather have avoided it, I've made the hinting threshold configurable in the font descriptor. That way, Arial's can be 40, while the CJK fonts enjoy 60.

Hence, tofu show up again as needed when Arial is used as a fallback font:

tofu

Personally, I'm very happy with the end result. Unless you have any more issues, I'll happily clean up the commit history so this can be merged.

File: src/openrct2/drawing/ttf_sdlport.c [hunk 9]
```diff
@@ -1272,6 +1433,20 @@ static bool CharacterIsDelimiter(char c, const char *delimiters)
     return false;
 }
 
+void TTF_SetFontHinting(TTF_Font* font, int hinting)
```
Comment (marijnvdwerf) [inline src/openrct2/drawing/ttf_sdlport.c]: This is only called with `0` or `1` for `hinting`.

File: src/openrct2/interface/Fonts.cpp [hunk 10]
```diff
@@ -26,38 +26,38 @@
 
 #ifndef NO_TTF
 TTFFontSetDescriptor TTFFontMSGothic = { {
-    { "msgothic.ttc", "MS PGothic", 9, 1, 0, 15, nullptr },
-    { "msgothic.ttc", "MS PGothic", 12, 1, 0, 17, nullptr },
-    { "msgothic.ttc", "MS PGothic", 12, 1, 0, 17, nullptr },
-    { "msgothic.ttc", "MS PGothic", 13, 1, 0, 20, nullptr },
+    { "msgothic.ttc", "MS PGothic", 9, 1, 0, 15, 60, nullptr },
```
Comment (marijnvdwerf) [inline src/openrct2/interface/Fonts.cpp]: Maybe add a define for these thresholds?

File: src/openrct2/drawing/ttf_sdlport.c (hunk 9 above)
Comment (AaronVanGeffen) [reply src/openrct2/drawing/ttf_sdlport.c]: Yes, but it is the original function backported from SDL_ttf. I figured that might be better to use, conceptually, rather than just hacking it in.

File: src/openrct2/interface/Fonts.cpp (hunk 10 above)
Comment (AaronVanGeffen) [reply src/openrct2/interface/Fonts.cpp]: Sure! Since this is C++, shall we implement them using `constexpr`? As for their names, what about `HINTING_THRESHOLD_LOW` for 40 and `HINTING_THRESHOLD_MEDIUM` for 60?

Comment (marijnvdwerf) [reply src/openrct2/interface/Fonts.cpp]: No idea what a `constexpr` is, but if it's better: sure :P. Names seem fine though.

Comment (AaronVanGeffen) [reply src/openrct2/interface/Fonts.cpp]: Actually, `constexpr` is probably overkill. I've just defined them as old-skool consts. :)

**Comment by marijnvdwerf** [comment]
Seems fine to me. Probably should revert the logo though.

**Comment by AaronVanGeffen** [comment]
Haha, yes, absolutely. Thanks. I'll do a rebase onto develop in the morning, squashing and dropping a commit here and there, then.

File: src/openrct2/interface/colour.c [hunk 11]
```diff
@@ -66,14 +66,14 @@ void colours_init_maps()
 #ifndef NO_TTF
 static uint8 BlendColourMap[PALETTE_COUNT][PALETTE_COUNT] = {0};
 
-static uint8 findClosestIndex(uint8 red, uint8 green, uint8 blue)
+static uint8 findClosestPaletteIndex(uint8 red, uint8 green, uint8 blue)
 {
     sint16 closest = -1;
-    sint32 closestDistance = INT32_MAX;
+    double closestDistance = INT32_MAX;
```
Comment (AaronVanGeffen) [inline src/openrct2/interface/colour.c]: @marijnvdwerf Do you reckon we should drop the doubles in favour of going back to sint32s? They don't seem to have made much of an impact.

Comment (marijnvdwerf) [reply src/openrct2/interface/colour.c]: I've just ran a test on all colours, int or double always produce the same outcome.

**Comment by AaronVanGeffen** [comment]
Rebased the lot on top of `develop`. Merged and dropped a few commits, too. Wondering whether I should squash 8c1ae5e651ebdb49a79e18e507e5da835aa60d86, 253b52573c00dd05c441b737c43b0df44b4cbb68, 899fe2faf967547416b669453a6248d080139961? Personally, I really like having the history around.

@marijnvdwerf, what do you think?

File: src/openrct2/interface/colour.h [hunk 12]
```diff
@@ -148,6 +151,10 @@ extern rct_colour_map ColourMapA[COLOUR_COUNT];
 
 void colours_init_maps();
 
+#ifndef NO_TTF
+uint8 blend(const uint8 paletteIndex1, const uint8 paletteIndex2);
```
Comment (marijnvdwerf) [inline src/openrct2/interface/colour.h]: Can you rename this to something more verbose?

**Comment by marijnvdwerf** [comment]
I'm fine with the history as-is. I'll merge it once you renamed the `blend` function 👍

**Comment by AaronVanGeffen** [comment]
Done! Thanks for the help and review!
